### PR TITLE
Add freeze_tensorrt and support loading frozen graphs in python

### PR DESCRIPTION
**Pull request by sethtroisi**
This is faster. thought it's not clear how much faster

### CODE REVIEW

**Comment by amj** [comment]
@sethtroisi what are we doing with this?  @chsigg mentioned he's got this in his batching stuff as well -- what's the plan here?

File: dual_net.py [hunk 1]
```diff
@@ -138,6 +139,12 @@
 # Per AGZ, 2048 minibatch * 1k = 2M positions/generation
 EXAMPLES_PER_GENERATION = 2 ** 21
 
+def load_graph_def(model):
```
Comment (amj) [inline dual_net.py]: needs comment

File: dual_net.py [hunk 2]
```diff
@@ -161,6 +172,29 @@ def initialize_graph(self):
             else:
                 self.sess.run(tf.global_variables_initializer())
 
+    def load_frozen_graph(self):
+        # Can either be a frozen graph or a frozen tensorrt graph.
+        graph_def = load_graph_def(self.save_file)
+
+        if self.save_file.endswith('.trt.pb'):
+            # NOTE: when using a TensorRT graph parallel-readouts should
```
Comment (amj) [inline dual_net.py]: should this note be a FATAL/abort with message?

File: dual_net.py [hunk 3]
```diff
@@ -161,6 +172,29 @@ def initialize_graph(self):
...
+            # match the hardcoded batch_size.
+            import tensorflow.contrib.tensorrt as trt
+
+        outputs = ['value_output', 'policy_output']
+        with self.sess.graph.as_default():
+            # TODO do I need to fix first dimension of features?
```
Comment (amj) [inline dual_net.py]: do you?

Comment (sethtroisi) [reply dual_net.py]: Turned into a TODO above

File: dual_net.py [hunk 4]
```diff
@@ -493,9 +527,28 @@ def export_model(model_path):
...
+
 def freeze_graph(model_path):
     n = DualNetwork(model_path)
     out_graph = tf.graph_util.convert_variables_to_constants(
         n.sess, n.sess.graph.as_graph_def(), ["policy_output", "value_output"])
     with tf.gfile.GFile(model_path + '.pb', 'wb') as f:
         f.write(out_graph.SerializeToString())
+
+
+def freeze_tensorrt(load_file, precision_mode='FP32', batch_size=8):
```
Comment (sethtroisi) [inline dual_net.py]: This isn't hooked up to CLI yet but that can be done later or as needed

File: dual_net.py [hunk 5]
```diff
@@ -138,7 +145,11 @@ def __init__(self, save_file):
         config = tf.ConfigProto()
         config.gpu_options.allow_growth = True
         self.sess = tf.Session(graph=tf.Graph(), config=config)
-        self.initialize_graph()
+
+        if save_file.endswith('.pb'):
```
[the remark this answers is not in the document]
Comment (sethtroisi) [reply dual_net.py]: Fixed

**Comment by chsigg** [comment]
The TensorRT backend in the cc player loads an .uff graph, so freezing the graph in the python code is done with the tf2uff converter, not the TensorRT integration of TensorFlow.

Here is what freezing a graph to uff looks like:
```
def freeze_graph(model_path):
    n = DualNetwork(model_path)
    outputs = ["policy_output", "value_output"]
    out_graph = tf.graph_util.convert_variables_to_constants(
        n.sess, n.sess.graph.as_graph_def(), outputs)
    with tf.gfile.GFile(model_path + '.pb', 'wb') as f:
        f.write(out_graph.SerializeToString())
    with tf.gfile.GFile(model_path + '.uff', 'wb') as f:
        f.write(uff.from_tensorflow(out_graph, outputs))
```

I was planning to just add that small modification as a patch to MiniGo when importing it for MLPerf. Or not even that and just let submitters figure it out themselves if they want to use TensorRT.

**Pull request closed without merging**